src/coreclr/tools/cdac-build-tool/DataDescriptorModel.cs

```diff
@@ -0,0 +1,446 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text.Json;
+
+namespace Microsoft.DotNet.Diagnostics.DataContract.BuildTool;
+
+public class DataDescriptorModel
+{
+    public string Baseline { get; }
+    public IReadOnlyDictionary<string, TypeModel> Types { get; }
+    public IReadOnlyDictionary<string, GlobalModel> Globals { get; }
+    public IReadOnlyDictionary<string, ContractModel> Contracts { get; }
+    public uint PlatformFlags { get; }
+    private DataDescriptorModel(string baseline, IReadOnlyDictionary<string, TypeModel> types, IReadOnlyDictionary<string, GlobalModel> globals, IReadOnlyDictionary<string, ContractModel> contracts, uint platformFlags)
+    {
+        Baseline = baseline;
+        Types = types;
+        Globals = globals;
+        Contracts = contracts;
+        PlatformFlags = platformFlags;
+    }
+
+    public const string PointerTypeName = "pointer";
+
+    internal void DumpModel()
+    {
+        Console.WriteLine("\nData Descriptor Model:");
+        Console.WriteLine($"Platform Flags: 0x{PlatformFlags:x8}");
+        Console.WriteLine($"Baseline: {Baseline}");
+        foreach (var (typeName, type) in Types)
+        {
+            Console.WriteLine($"Type: {typeName}");
+            if (type.Size != null)
+            {
+                Console.WriteLine($"  Size: 0x{type.Size:x8}");
+            }
+            foreach (var (fieldName, field) in type.Fields)
+            {
+                Console.WriteLine($"  Field: {fieldName}");
+                Console.WriteLine($"    Type: {field.Type}");
+                Console.WriteLine($"    Offset: 0x{field.Offset:x8}");
+            }
+        }
+        foreach (var (globalName, global) in Globals)
+        {
+            Console.WriteLine($"Global: {globalName}");
+            Console.WriteLine($"  Type: {global.Type}");
+            Console.WriteLine($"  Value: {global.Value}");
+        }
+        foreach (var (contractName, contract) in Contracts)
+        {
+            Console.WriteLine($"Contract: {contractName}");
+            Console.WriteLine($"  Version: {contract.Version}");
+        }
+    }
+
+    public string ToJson()
```

Comment:
Because we're doing things in a slightly weird way for the "compact" format, we will need custom serializers. At that point it's just more boilerplate to get to this same place, imo.